In [2]:
from datasets import load_dataset

/Users/shreeyadhakal/Desktop/Projects-2023/venv/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
# !pip install datasets

### Aya Dataset

***Description from [HuggingFace](https://huggingface.co/datasets/CohereForAI/aya_dataset):***
    The Aya Dataset is a multilingual instruction fine-tuning dataset curated by an open-science community via Aya Annotation Platform from Cohere For AI. The dataset contains a total of 204k human-annotated prompt-completion pairs along with the demographics data of the annotators.

In [54]:
aya_dataset = load_dataset("CohereForAI/aya_dataset")
aya_dataset

DatasetDict({
    train: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 202362
    })
    test: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 1750
    })
})

In [55]:
aya_nepali = aya_dataset.filter(lambda example: example['language'] == "Nepali")
aya_nepali

DatasetDict({
    train: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 4002
    })
    test: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 0
    })
})

In [56]:
print(aya_nepali['train']['inputs'][1])

दिइएको कवितांश पढ्नुहोस्‌ र सङ्क्षिप्त उत्तर लेख्नुहोस्‌ :

डिगर्चामा डोब तिम्रो चिसो हिउँभित्र होला वेत्रावती किनारभरि पौरखको
चिनो होला वीर पुर्खा! तिमीलाई मितेरीले मात्र बाँध्यो सागर तरी
संसारभरि वीर गोर्खा रगत बग्यो।

प्रश्न : हामीले हाम्रा पुर्खाको गौरव कसरी जोगाउन सक्छौं ?



In [57]:
print(aya_nepali['train']['targets'][1])


उत्तरः पुर्खाले गरेको त्याग, तपस्या र बलिदानको उच्च सम्मान गर्दै एवम्‌
देशभक्तिपूर्ण बाटोलाई तन, मन, वचन तथा कर्मले आत्मसात्‌ गरी हाम्रा
पुर्खाको गौरव जोगाउन सक्छौं ।




In [62]:
empty_text = aya_nepali.filter(lambda example: example['inputs'] == "" or example['targets'] == "")
empty_text

Filter: 100%|██████████| 4002/4002 [00:00<00:00, 22154.56 examples/s]


DatasetDict({
    train: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 0
    })
    test: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 0
    })
})

In [70]:
avg_inputs_len = sum(len(text.split()) for text in aya_nepali['train']['inputs']) / len(aya_nepali['train']['inputs'])
avg_outputs_len = sum(len(text.split()) for text in aya_nepali['train']['targets']) / len(aya_nepali['train']['targets'])
(avg_inputs_len, avg_outputs_len)

(14.914042978510745, 40.942278860569715)

In [64]:
aya_hindi = aya_dataset.filter(lambda example: example['language'] == "Hindi")
aya_hindi

DatasetDict({
    train: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 1153
    })
    test: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 0
    })
})

In [65]:
print(aya_hindi['train']['inputs'][1])

दाल मखनी कैसे बनाते हैं?


In [66]:
print(aya_hindi['train']['targets'][1])

दाल मखनी बनाने के लिए सबसे पहले काली उरद दाल और राजमा को रात भर भिगो दें। फिर इसे प्रेशर कुकर में पकाएं। एक पैन में मक्खन गरम करें, जीरा, अदरक-लहसुन का पेस्ट, प्याज, टमाटर, मिर्च पाउडर, और गरम मसाला डालकर भुनें। पकी हुई दाल और राजमा को इसमें मिलाएं और धीमी आंच पर पकने दें। क्रीम मिलाकर गाढ़ा होने तक पकाएं।


In [67]:
empty_text = aya_hindi.filter(lambda example: example['inputs'] == "" or example['targets'] == "")
empty_text

Filter:   0%|          | 0/1153 [00:00<?, ? examples/s]

Filter: 100%|██████████| 1153/1153 [00:00<00:00, 18545.56 examples/s]


DatasetDict({
    train: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 0
    })
    test: Dataset({
        features: ['inputs', 'targets', 'language', 'language_code', 'annotation_type', 'user_id'],
        num_rows: 0
    })
})

In [71]:
avg_inputs_len = sum(len(text.split()) for text in aya_hindi['train']['inputs']) / len(aya_hindi['train']['inputs'])
avg_outputs_len = sum(len(text.split()) for text in aya_hindi['train']['targets']) / len(aya_hindi['train']['targets'])
(avg_inputs_len, avg_outputs_len)

(23.617519514310494, 80.50910667823071)

There are only 4002 and 1153 samples for Nepali and Hindi, respectively. We can do few-shot experiments with those. We will have to see if it is enough for LoRA. Else, we will use Alpaca datasets translated into Nepali and Hindi.

### Alpaca Dataset

Alpaca Dataset is 52K instruction-following data by Stanford. We will be using the translated version of [Alpaca Clean](https://huggingface.co/datasets/yahma/alpaca-cleaned): 
- [Nepali](https://huggingface.co/datasets/Saugatkafley/alpaca-nepali-sft)
- [Hindi](https://huggingface.co/datasets/iamshnoo/alpaca-cleaned-hindi) - Translated from yahma/alpaca-cleaned using NLLB-1.3B


In [43]:
alpaca_nepali = load_dataset("Saugatkafley/alpaca-nepali-sft")
alpaca_nepali

Generating train split: 100%|██████████| 52005/52005 [00:00<00:00, 327891.81 examples/s]


DatasetDict({
    train: Dataset({
        features: ['instruction', 'input', 'output', 'id'],
        num_rows: 52005
    })
})

In [45]:
print(alpaca_nepali['train']['instruction'][1])

तीन प्राथमिक रंगहरू के हुन्?


In [46]:
print(alpaca_nepali['train']['input'][1])

In [47]:
print(alpaca_nepali['train']['output'][1])

तीन प्राथमिक रंगहरू रातो, नीलो र पहेंलो हुन्।


In [73]:
empty_text = alpaca_nepali.filter(lambda example: example['input'] == "" or example['output'] == "" or example['instruction'] == "")
empty_text

Filter:   0%|          | 0/52005 [00:00<?, ? examples/s]

Filter: 100%|██████████| 52005/52005 [00:00<00:00, 79630.14 examples/s]


DatasetDict({
    train: Dataset({
        features: ['instruction', 'input', 'output', 'id'],
        num_rows: 31337
    })
})

In [126]:
empty_text = alpaca_nepali.filter(lambda example: example['instruction'] is None)
empty_text

Filter: 100%|██████████| 52005/52005 [00:00<00:00, 91364.42 examples/s]


DatasetDict({
    train: Dataset({
        features: ['instruction', 'input', 'output', 'id'],
        num_rows: 3
    })
})

In [95]:
empty_text = alpaca_nepali.filter(lambda example: example['instruction'] == "" or example['instruction'] is None)
empty_text

Filter:   0%|          | 0/52005 [00:00<?, ? examples/s]

Filter: 100%|██████████| 52005/52005 [00:00<00:00, 76304.33 examples/s]


DatasetDict({
    train: Dataset({
        features: ['instruction', 'input', 'output', 'id'],
        num_rows: 3
    })
})

In [124]:
empty_text = alpaca_nepali.filter(lambda example: example['output'] == "" or example['output'] is None)
empty_text

DatasetDict({
    train: Dataset({
        features: ['instruction', 'input', 'output', 'id'],
        num_rows: 42
    })
})

In [122]:
alpaca_nepali_filtered = alpaca_nepali.filter(lambda example: example['output'] is not None and example['output'] != "")
alpaca_nepali_filtered = alpaca_nepali_filtered.filter(lambda example: example['instruction'] != "" and example['instruction'] is not None)
alpaca_nepali_filtered = alpaca_nepali_filtered.filter(lambda example: example['input'] is not None)
alpaca_nepali_filtered

Filter:   0%|          | 0/51963 [00:00<?, ? examples/s]

Filter: 100%|██████████| 51963/51963 [00:01<00:00, 40964.07 examples/s]


DatasetDict({
    train: Dataset({
        features: ['instruction', 'input', 'output', 'id'],
        num_rows: 51959
    })
})

In [125]:
avg_inst_len = sum(len(text.split()) for text in alpaca_nepali_filtered['train']['instruction']) / len(alpaca_nepali_filtered['train']['instruction'])
avg_inputs_len = sum(len(text.split()) for text in alpaca_nepali_filtered['train']['input']) / len(alpaca_nepali_filtered['train']['input'])
avg_outputs_len = sum(len(text.split()) for text in alpaca_nepali_filtered['train']['output']) / len(alpaca_nepali_filtered['train']['output'])
(avg_inst_len, avg_inputs_len, avg_outputs_len)

(8.68538655478358, 3.4686579803306454, 38.96822494659251)

In [42]:
alpaca_hindi = load_dataset("iamshnoo/alpaca-cleaned-hindi")
alpaca_hindi

Generating train split: 100%|██████████| 51760/51760 [00:00<00:00, 290897.61 examples/s]


DatasetDict({
    train: Dataset({
        features: ['input', 'instruction', 'output'],
        num_rows: 51760
    })
})

In [51]:
print(alpaca_hindi['train']['instruction'][1])

तीन प्राथमिक रंग क्या हैं?


In [49]:
print(alpaca_hindi['train']['input'][1])

In [50]:
print(alpaca_hindi['train']['output'][1])

प्रकाश के लिए उपयोग किए जाने वाले योजक रंग प्रणाली में, प्राथमिक रंग लाल, हरे और नीले (आरजीबी) हैं। प्रकाश के लिए उपयोग किए जाने वाले योजक रंग प्रणाली में, प्राथमिक रंग लाल, हरे और नीले (आरजीबी) हैं।


In [129]:
empty_text = alpaca_hindi.filter(lambda example: example['input'] is None)
empty_text

Filter:   0%|          | 0/51760 [00:00<?, ? examples/s]

Filter: 100%|██████████| 51760/51760 [00:01<00:00, 49712.86 examples/s]


DatasetDict({
    train: Dataset({
        features: ['input', 'instruction', 'output'],
        num_rows: 0
    })
})

In [132]:
empty_text = alpaca_hindi.filter(lambda example: example['instruction'] == "" or example['instruction'] is None)
empty_text

Filter: 100%|██████████| 51760/51760 [00:00<00:00, 64501.34 examples/s]


DatasetDict({
    train: Dataset({
        features: ['input', 'instruction', 'output'],
        num_rows: 0
    })
})

In [131]:
empty_text = alpaca_hindi.filter(lambda example: example['output'] == "" or example['output'] is None)
empty_text

Filter: 100%|██████████| 51760/51760 [00:00<00:00, 84864.02 examples/s]


DatasetDict({
    train: Dataset({
        features: ['input', 'instruction', 'output'],
        num_rows: 0
    })
})

In [133]:
avg_inst_len = sum(len(text.split()) for text in alpaca_hindi['train']['instruction']) / len(alpaca_hindi['train']['instruction'])
avg_inputs_len = sum(len(text.split()) for text in alpaca_hindi['train']['input']) / len(alpaca_hindi['train']['input'])
avg_outputs_len = sum(len(text.split()) for text in alpaca_hindi['train']['output']) / len(alpaca_hindi['train']['output'])
(avg_inst_len, avg_inputs_len, avg_outputs_len)

(11.920595054095827, 4.6264296754250385, 112.07911514683153)